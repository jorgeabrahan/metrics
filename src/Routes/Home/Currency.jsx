import { NavLink } from 'react-router-dom';
import PropTypes from 'prop-types';

const Currency = ({
  short, name,
}) => (
  <NavLink to={`/details/${short}`}>
    <div>
      <h2 className="currency__title">
        { name }
      </h2>
      <small>{ short }</small>
    </div>
    <div className="currency__icon">
      <span className="material-symbols-outlined">
        arrow_right_alt
      </span>
    </div>
  </NavLink>
);

Currency.propTypes = {
  short: PropTypes.string.isRequired,
  name: PropTypes.string.isRequired,
};

export default Currency;
