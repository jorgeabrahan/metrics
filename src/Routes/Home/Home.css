.currencies__header {
  background-image: url('../../currency.svg');
  background-repeat: no-repeat;
  background-size: 220px;
  background-position: 0.5rem center;

  height: 25vh;
  padding: 0.5rem;
  
  display: flex;
  flex-direction: column;
  align-items: flex-end;
  justify-content: center;
}

.currencies__header h2 {
  text-align: right;
  font-weight: 800;
  font-size: 1.4rem;
  line-height: 1.7rem;
  text-transform: uppercase;
}

.currencies__header small {
  display: block;
  text-align: right;
}

.currencies__title {
  background-color: var(--ruber);
  display: block;
  width: 100%;
  font-weight: bold;
  padding: 0.5rem 1.5rem;
  text-transform: uppercase;
}

.currencies a {
  text-decoration: none;
  color: white;
  display: flex;
  align-items: center;
  justify-content: space-between;
  padding: 1rem 0.5rem;
}

.currency small {
  display: block;
}

.currencies a:nth-child(even) {
  background-color: var(--ruber);
}
.currencies a:nth-child(odd) {
  background-color: var(--dark-pink);
}

.currency__title {
  line-height: 1.5rem;
}

.currency__icon {
  display: flex;
  width: max-content;
  padding: 0.1rem;
  border-radius: 50%;
  border: 2px solid white;
}

.currency__icon span {
  font-size: 1rem;
}