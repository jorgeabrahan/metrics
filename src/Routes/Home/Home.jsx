/* eslint-disable import/no-extraneous-dependencies */
import { useEffect, useMemo } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { v4 as uuid } from 'uuid';
import { fetchCurrencies, searchForCurrency } from '../../redux/Currencies/currenciesSlice';
import { setQuery } from '../../redux/Search/searchSlice';
import Currency from './Currency';
import './Home.css';

const createList = (currencies) => currencies.map(({ short, name }) => (
  <Currency
    key={uuid()}
    short={short}
    name={name}
  />
));

const Home = () => {
  const dispatch = useDispatch();
  const {
    currencies, filteredCurrencies, status, error,
  } = useSelector((store) => store.currencies);
  const { isSearching } = useSelector((store) => store.search);

  useEffect(() => {
    if (currencies.length !== 0) return;
    dispatch(fetchCurrencies());
  }, [dispatch]);

  useEffect(() => {
    if (isSearching) return;
    dispatch(searchForCurrency(''));
    dispatch(setQuery(''));
  }, [dispatch, isSearching]);

  const filteredCurrenciesList = useMemo(() => (
    createList(filteredCurrencies)
  ), [filteredCurrencies]);

  let render = <p>Loading...</p>;
  if (status === 'fulfilled') render = filteredCurrenciesList;
  if (status === 'rejected') {
    render = (
      <p>
        [Error]:
        {error}
      </p>
    );
  }

  return (
    <section>
      <div className="header">
        <h2>
          Currency
          <br />
          Rates
        </h2>
        <small>Includes cryptocurrencies</small>
      </div>
      <small className="subtitle-separator">List of available currencies</small>
      <div className="currencies">
        {render}
      </div>
    </section>
  );
};

export default Home;
