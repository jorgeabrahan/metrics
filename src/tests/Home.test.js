/* eslint-disable import/no-extraneous-dependencies */
import { render, screen } from '@testing-library/react';
import renderer from 'react-test-renderer';
import axios from 'axios';
import Home from '../Routes/Home/Home';
import wrap from './testResources';

jest.mock('axios');

axios.get.mockResolvedValue({
  data: {
    afn: 'Afghan afghani',
    algo: 'Algorand',
    all: 'Albanian lek',
    amd: 'Armenian dram',
  },
});

describe('Tests for <Home /> component', () => {
  render(wrap(<Home />));
  it('Should render childs from API correctly', () => {
    expect(screen.getByText('Afghan afghani')).not.toBeNull();
    expect(screen.getByText('Algorand')).not.toBeNull();
    expect(screen.getByText('Albanian lek')).not.toBeNull();
    expect(screen.getByText('Armenian dram')).not.toBeNull();
  });

  it('Should not display the loading message', () => {
    expect(() => screen.getByText('Loading...')).toThrow();
  });

  it('Should display currencies as links', () => {
    render(wrap(<Home />));
    expect(screen.getByText('Afghan afghani').closest('a').href.includes('afn')).toBeTruthy();
    expect(screen.getByText('Algorand').closest('a').href.includes('algo')).toBeTruthy();
    expect(screen.getByText('Albanian lek').closest('a').href.includes('all')).toBeTruthy();
    expect(screen.getByText('Armenian dram').closest('a').href.includes('amd')).toBeTruthy();
  });

  it('Should match the snapshot', () => {
    expect(renderer.create(wrap(<Home />)).toJSON()).toMatchSnapshot();
  });
});
