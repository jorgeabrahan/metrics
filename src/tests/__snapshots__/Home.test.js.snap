// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tests for <Home /> component Should match the snapshot 1`] = `
<section>
  <div
    className="header"
  >
    <h2>
      Currency
      <br />
      Rates
    </h2>
    <small>
      Includes cryptocurrencies
    </small>
  </div>
  <small
    className="subtitle-separator"
  >
    List of available currencies
  </small>
  <div
    className="currencies"
  >
    <a
      className=""
      href="/details/afn"
      onClick={[Function]}
    >
      <div>
        <h2
          className="currency__title"
        >
          Afghan afghani
        </h2>
        <small>
          afn
        </small>
      </div>
      <div
        className="currency__icon"
      >
        <span
          className="material-symbols-outlined"
        >
          arrow_right_alt
        </span>
      </div>
    </a>
    <a
      className=""
      href="/details/algo"
      onClick={[Function]}
    >
      <div>
        <h2
          className="currency__title"
        >
          Algorand
        </h2>
        <small>
          algo
        </small>
      </div>
      <div
        className="currency__icon"
      >
        <span
          className="material-symbols-outlined"
        >
          arrow_right_alt
        </span>
      </div>
    </a>
    <a
      className=""
      href="/details/all"
      onClick={[Function]}
    >
      <div>
        <h2
          className="currency__title"
        >
          Albanian lek
        </h2>
        <small>
          all
        </small>
      </div>
      <div
        className="currency__icon"
      >
        <span
          className="material-symbols-outlined"
        >
          arrow_right_alt
        </span>
      </div>
    </a>
    <a
      className=""
      href="/details/amd"
      onClick={[Function]}
    >
      <div>
        <h2
          className="currency__title"
        >
          Armenian dram
        </h2>
        <small>
          amd
        </small>
      </div>
      <div
        className="currency__icon"
      >
        <span
          className="material-symbols-outlined"
        >
          arrow_right_alt
        </span>
      </div>
    </a>
  </div>
</section>
`;
